pkg/controller/node/node_controller.go

```diff
@@ -521,43 +516,49 @@ func (ctrl *Controller) setDesiredMachineConfigAnnotation(nodeName, currentConfi
 	})
 }
 
-func makeProgress(pool *mcfgv1.MachineConfigPool, nodes []*corev1.Node) (int, error) {
-	maxunavail, err := maxUnavailable(pool, nodes)
-	if err != nil {
-		return 0, err
-	}
-	unavail := len(getUnavailableMachines(pool.Status.Configuration.Name, nodes))
-	progress := 0
-	if unavail < maxunavail {
-		progress = maxunavail - unavail
-	}
-
-	return progress, nil
-}
+func getCandidateMachines(pool *mcfgv1.MachineConfigPool, nodesInPool []*corev1.Node, maxUnavailable int) []*corev1.Node {
+	targetConfig := pool.Status.Configuration.Name
 
-func getCandidateMachines(pool *mcfgv1.MachineConfigPool, nodes []*corev1.Node, progress int) []*corev1.Node {
-	acted := getReadyMachines(pool.Status.Configuration.Name, nodes)
-	acted = append(acted, getUnavailableMachines(pool.Status.Configuration.Name, nodes)...)
+	unavail := getUnavailableMachines(nodesInPool)
+	// If we're at capacity, there's nothing to do.
+	if len(unavail) >= maxUnavailable {
+		return nil
+	}
+	capacity := maxUnavailable - len(unavail)
+	failingThisConfig := 0
+	// We only look at nodes which aren't already targeting our desired config
+	var nodes []*corev1.Node
+	for _, node := range nodesInPool {
+		if node.Annotations[daemonconsts.DesiredMachineConfigAnnotationKey] == targetConfig {
+			if isNodeMCDFailing(node) {
+				failingThisConfig++
+			}
+			continue
+		}
 
-	actedMap := map[string]bool{}
-	for _, node := range acted {
-		actedMap[node.Name] = true
+		nodes = append(nodes, node)
 	}
 
-	var candidates []*corev1.Node
-	for _, node := range nodes {
-		if !actedMap[node.Name] {
-			candidates = append(candidates, node)
-		}
+	// Nodes which are failing to target this config also count against
+	// availability - it might be a transient issue, and if the issue
+	// clears we don't want multiple to update at once.
+	if failingThisConfig >= capacity {
+		return nil
 	}
+	capacity -= failingThisConfig
 
-	if len(candidates) <= progress {
-		return candidates
+	if len(nodes) < capacity {
+		return nodes
 	}
-	return candidates[:progress]
+	return nodes[:capacity]
 }
 
 func maxUnavailable(pool *mcfgv1.MachineConfigPool, nodes []*corev1.Node) (int, error) {
+	// Hardcoded for now per discussion in
+	// https://github.com/openshift/machine-config-operator/pull/701
+	if pool.Name == "master" {
+		return 1, nil
+	}
```

Comment:
just triple checking with @RobertKrawitz and @hexfusion that we never wants more than 1 master unschedulable/rebooted